sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveTableScanSuite.scala

```diff
@@ -187,6 +189,60 @@ class HiveTableScanSuite extends HiveComparisonTest with SQLTestUtils with TestH
     }
   }
 
+  test("SPARK-28169: Convert scan predicate condition to CNF") {
+    withTable("t", "temp") {
+      sql(
+        s"""
+           |CREATE TABLE t(i int)
+           |PARTITIONED BY (p int)
+           |STORED AS textfile""".stripMargin)
+      spark.range(0, 1000, 1).selectExpr("id as col")
+        .createOrReplaceTempView("temp")
+
+      for (part <- Seq(1, 2, 3, 4)) {
+        sql(
+          s"""
+             |INSERT OVERWRITE TABLE t PARTITION (p='$part')
+             |select col from temp""".stripMargin)
+      }
+
+      val scan1 = getHiveTableScanExec(
+        "SELECT * FROM t WHERE p = '1' OR (p = '2' AND i = 1)")
+      val scan2 = getHiveTableScanExec(
+        "SELECT * FROM t WHERE (p = '1' and i = 2) or (i = 1 or p = '2')")
+      val scan3 = getHiveTableScanExec(
+        "SELECT * FROM t WHERE (p = '1' and i = 2) or (p = '3' and i = 3 )")
+      val scan4 = getHiveTableScanExec(
+        "SELECT * FROM t WHERE (p = '1' and i = 2) or (p = '2' or p = '3')")
+      val scan5 = getHiveTableScanExec(
+        "SELECT * FROM t")
+      val scan6 = getHiveTableScanExec(
+        "SELECT * FROM t where p = '1' and i = 2")
+
+      val scan7 = getHiveTableScanExec(
+        """
+          |SELECT i, COUNT(1) FROM (
+          |SELECT * FROM t where  p = '1' OR (p = '2' AND i = 1)
+          |) TMP GROUP BY i
+        """.stripMargin)
+
+      assert(scan1.prunedPartitions.map(_.toString) ==
+        Stream("t(p=1)", "t(p=2)"))
```

Comment:
It's better to put the query and the check together.

Reply:
> It's better to put the query and the check together.

DOne